### PR TITLE
badge_viewed

### PR DESCRIPTION
**Pull request by invaliduser**
`badge_viewed` event

### CODE REVIEW

File: src/transformer/events/core/badge_viewed.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    $user=$repo->read_record_by_id('user',$event->userid); 
+    $course = (isset($event->courseid) && $event->courseid != 0)
+        ? $repo->read_record_by_id('course', $event->courseid)
+        : null;
+    $lang = is_null($course) ? $config['source_lang'] : utils\get_course_lang($course);
+
+    $statement = [
+        'actor' => utils\get_user($config,$user),
+        'verb' => ['id' => 'http://id.tincanapi.com/verb/viewed',
+                   'display' => ['en' => 'Viewed']
```
Comment (milt) [inline src/transformer/events/core/badge_viewed.php:57]: should not be hardcoded

File: tests/core/badge_viewed/user_viewed_badge/data.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    }
+  ],
+  "course": [
+    {
+      "fullname": "test_course_name",
+      "id": 1,
+      "lang": "en",
+      "summary": "test_course_summary"
+    }
+  ],
```
Comment (milt) [inline tests/core/badge_viewed/user_viewed_badge/data.json]: not needed

File: tests/core/badge_viewed/user_viewed_badge/data.json [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  "user": [
+    {
+      "email": "viewer@test.com",
+      "firstname": "badge_viewer_firstname",
+      "id": 1,
+      "lastname": "badge_viewer_lastname"
+    }
+  ]
+}
```
Comment (milt) [inline tests/core/badge_viewed/user_viewed_badge/data.json]: not needed

File: tests/core/badge_viewed/user_viewed_badge/user_viewed_badge_test.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+     * @return string
+     */
+    protected function get_plugin_name() {
+        return "core";
+    }
+
+    /**
+     * Appease auto-detecting of test cases. xapi_test_case has default test cases.
+     *
+     * @covers ::attempt_submitted
```
Comment (milt) [inline tests/core/badge_viewed/user_viewed_badge/user_viewed_badge_test.php]: don't forget the annotation

File: src/transformer/events/core/badge_viewed.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+          'definition' => [
+            'name' => [$lang =>$badge->name],
+            'description' => [$lang => $badge->description],
+            'type' =>   'https://xapi.edlm/profiles/edlm-lms/concepts/activity-types/badge',
+            'extensions' => [
+              'https://xapi.edlm/profiles/edlm-lms/v1/concepts/activity-extensions/badge-type' =>  $badgetype,
+              'https://xapi.edlm/profiles/edlm-lms/v1/concepts/activity-extensions/badge-version' => $badge->version
+            ]
+          ],
+        ],
```
Comment (milt) [inline src/transformer/events/core/badge_viewed.php]: is there not a function for making badge objects? If not, make one so we are consistent.

**Review by milt — CHANGES_REQUESTED** [review]
See various

File: tests/core/badge_viewed/user_viewed_badge/data.json (hunk 2 above)
Comment (invaliduser) [reply tests/core/badge_viewed/user_viewed_badge/data.json]: removed

File: tests/core/badge_viewed/user_viewed_badge/data.json (hunk 3 above)
Comment (invaliduser) [reply tests/core/badge_viewed/user_viewed_badge/data.json]: removed

File: tests/core/badge_viewed/user_viewed_badge/user_viewed_badge_test.php (hunk 4 above)
Comment (invaliduser) [reply tests/core/badge_viewed/user_viewed_badge/user_viewed_badge_test.php]: amended

File: src/transformer/events/core/badge_viewed.php (hunk 5 above)
Comment (invaliduser) [reply src/transformer/events/core/badge_viewed.php]: added [badge_object](https://github.com/yetanalytics/moodle-logstore_xapi/pull/55/files#diff-491877f84c165e46636c9e57ac33ded9bb5d3c90b485285b0c73e71c28782018) function

File: src/transformer/events/core/badge_revoked.php [hunk 6]
```diff
@@ -41,7 +41,6 @@ function badge_revoked(array $config, \stdClass $event) {
     $revoker = utils\get_user($config, $repo->read_record_by_id('user', $event->userid));
     $course = $badge->courseid ? $repo->read_record_by_id('course', $badge->courseid) : null;
     $lang = $badge->language ?? 'en';
```
Comment (milt) [inline src/transformer/events/core/badge_revoked.php]: maybe fall back to the course lang after the badge lang, then after that config source lang

File: src/transformer/events/core/badge_viewed.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    global $CFG;
+    $repo = $config['repo'];
+    $badge = $repo->read_record_by_id('badge', $event->objectid);
+
+    //all three here may not exist
+    $user=$repo->read_record_by_id('user',$event->userid); 
+    $course = (isset($event->courseid) && $event->courseid != 0)
+        ? $repo->read_record_by_id('course', $event->courseid)
+        : null;
+    $lang = is_null($course) ? $config['source_lang'] : utils\get_course_lang($course);
```
Comment (milt) [inline src/transformer/events/core/badge_viewed.php]: wouldn't you want to be the lang to be consistent across different badge events?

**Review by milt — CHANGES_REQUESTED** [review]
Looks good except for the differing strategy for lang, this should be consistent. As mentioned elsewhere I think the order is:

1. badge lang (if set)
2. course lang (if there is a course)
3. config source lang

File: src/transformer/events/core/badge_viewed.php (hunk 7 above)
Comment (invaliduser) [reply src/transformer/events/core/badge_viewed.php]: changed

File: src/transformer/events/core/badge_revoked.php (hunk 6 above)
Comment (milt) [reply src/transformer/events/core/badge_revoked.php]: this is still not updated

Comment (invaliduser) [reply src/transformer/events/core/badge_revoked.php]: changed to be consistent

**Review by milt — APPROVED** [review]

**Pull request merged**